File: sdk/typescript/src/schema.ts
/** Harness-side JSON Schema subset used for outputSchema. Unsupported keywords are rejected. */

const ALLOWED = new Set([
  "type", "properties", "required", "items", "enum", "additionalProperties",
  "minLength", "maxLength", "minimum", "maximum", "minItems", "maxItems",
  "description", "title", "default",
]);
const JSON_FENCE = /```(?:json)?\s*(\{[\s\S]*?\}|\[[\s\S]*?\])\s*```/;

export function assertSupported(schema: unknown): void {
  if (!schema || typeof schema !== "object" || Array.isArray(schema) || !Object.keys(schema).length) {
    throw new Error("outputSchema must be a non-empty object");
  }
  walk(schema);
}

function walk(node: unknown): void {
  if (!node || typeof node !== "object" || Array.isArray(node)) return;
  const record = node as Record<string, unknown>;
  if ("$ref" in record || "$id" in record || "$schema" in record) {
    throw new Error("outputSchema must not use $ref, $id, or $schema");
  }
  for (const key of Object.keys(record)) {
    if (!ALLOWED.has(key)) throw new Error(`outputSchema has unsupported keyword: ${key}`);
  }
  if (record.properties && typeof record.properties === "object") {
    for (const child of Object.values(record.properties as Record<string, unknown>)) walk(child);
  }
  if (record.items && typeof record.items === "object") walk(record.items);
}

export function extractJson(text: string): unknown {
  let raw = (text || "").trim();
  if (!raw) throw new Error("final text is empty");
  const fenced = JSON_FENCE.exec(raw);
  if (fenced) raw = fenced[1];
  else {
    const obj = raw.indexOf("{");
    const arr = raw.indexOf("[");
    const starts = [obj, arr].filter((index) => index >= 0);
    if (starts.length) raw = raw.slice(Math.min(...starts));
  }
  return JSON.parse(raw);
}

export function validate(value: unknown, schema: Record<string, unknown>, path = "$"): string[] {
  const errors: string[] = [];
  const expected = schema.type;
  if (expected) {
    const types = Array.isArray(expected) ? expected : [expected];
    if (!types.some((kind) => isType(value, String(kind)))) {
      errors.push(`${path} should be ${expected as string}`);
      return errors;
    }
  }
  if ("enum" in schema && !((schema.enum as unknown[]) || []).includes(value)) {
    errors.push(`${path} is not one of the allowed values`);
  }
  if (typeof value === "string") {
    if (typeof schema.minLength === "number" && value.length < schema.minLength) {
      errors.push(`${path} is shorter than minLength`);
    }
    if (typeof schema.maxLength === "number" && value.length > schema.maxLength) {
      errors.push(`${path} is longer than maxLength`);
    }
  }
  if (typeof value === "number" && !Number.isNaN(value)) {
    if (typeof schema.minimum === "number" && value < schema.minimum) errors.push(`${path} is below minimum`);
    if (typeof schema.maximum === "number" && value > schema.maximum) errors.push(`${path} is above maximum`);
  }
  if (Array.isArray(value)) {
    if (typeof schema.minItems === "number" && value.length < schema.minItems) errors.push(`${path} has too few items`);
    if (typeof schema.maxItems === "number" && value.length > schema.maxItems) errors.push(`${path} has too many items`);
    if (schema.items && typeof schema.items === "object" && !Array.isArray(schema.items)) {
      value.forEach((item, index) => {
        errors.push(...validate(item, schema.items as Record<string, unknown>, `${path}[${index}]`));
      });
    }
  }
  if (value && typeof value === "object" && !Array.isArray(value)) {
    const record = value as Record<string, unknown>;
    const props = (schema.properties && typeof schema.properties === "object")
      ? schema.properties as Record<string, Record<string, unknown>> : {};
    const required = Array.isArray(schema.required) ? schema.required as string[] : [];
    for (const key of required) {
      if (!(key in record)) errors.push(`${path}.${key} is required`);
    }
    const additional = schema.additionalProperties === undefined ? true : schema.additionalProperties;
    for (const [key, item] of Object.entries(record)) {
      if (key in props) errors.push(...validate(item, props[key], `${path}.${key}`));
      else if (additional === false) errors.push(`${path}.${key} is not allowed`);
      else if (additional && typeof additional === "object") {
        errors.push(...validate(item, additional as Record<string, unknown>, `${path}.${key}`));
      }
    }
  }
  return errors;
}

function isType(value: unknown, kind: string): boolean {
  if (kind === "object") return Boolean(value) && typeof value === "object" && !Array.isArray(value);
  if (kind === "array") return Array.isArray(value);
  if (kind === "string") return typeof value === "string";
  if (kind === "integer") return typeof value === "number" && Number.isInteger(value);
  if (kind === "number") return typeof value === "number" && !Number.isNaN(value);
  if (kind === "boolean") return typeof value === "boolean";
  if (kind === "null") return value === null;
  return false;
}

File: sdk/python/dgc_sdk/schema.py
"""Harness-side JSON Schema subset used for ``output_schema``.

This is not a full JSON Schema implementation. Unsupported keywords are rejected rather than
ignored. Remote ``$ref`` is refused. Repair happens in the session after tools have finished.
"""

from __future__ import annotations

import json
import re
from typing import Any, Mapping

from .errors import DGCConfigError

_ALLOWED = frozenset({
    "type", "properties", "required", "items", "enum", "additionalProperties",
    "minLength", "maxLength", "minimum", "maximum", "minItems", "maxItems",
    "description", "title", "default",
})
_JSON_FENCE = re.compile(r"```(?:json)?\s*(\{.*?\}|\[.*?\])\s*```", re.DOTALL)


def assert_supported(schema: Mapping[str, Any]) -> None:
    if not isinstance(schema, dict) or not schema:
        raise DGCConfigError("output_schema must be a non-empty object")
    _walk(schema)


def _walk(node: Any) -> None:
    if not isinstance(node, dict):
        return
    if "$ref" in node or "$id" in node or "$schema" in node:
        raise DGCConfigError("output_schema must not use $ref, $id, or $schema")
    unknown = set(node) - _ALLOWED
    if unknown:
        raise DGCConfigError(f"output_schema has unsupported keyword: {sorted(unknown)[0]}")
    props = node.get("properties")
    if isinstance(props, dict):
        for child in props.values():
            _walk(child)
    items = node.get("items")
    if isinstance(items, dict):
        _walk(items)


def extract_json(text: str) -> Any:
    raw = (text or "").strip()
    if not raw:
        raise ValueError("final text is empty")
    match = _JSON_FENCE.search(raw)
    if match:
        raw = match.group(1)
    else:
        start_obj, start_arr = raw.find("{"), raw.find("[")
        starts = [index for index in (start_obj, start_arr) if index >= 0]
        if starts:
            raw = raw[min(starts):]
    return json.loads(raw)


def validate(value: Any, schema: Mapping[str, Any], path: str = "$") -> list[str]:
    errors: list[str] = []
    expected = schema.get("type")
    if expected:
        types = expected if isinstance(expected, list) else [expected]
        if not any(_is_type(value, kind) for kind in types):
            errors.append(f"{path} should be {expected}")
            return errors
    if "enum" in schema and value not in schema["enum"]:
        errors.append(f"{path} is not one of the allowed values")
    if isinstance(value, str):
        if "minLength" in schema and len(value) < int(schema["minLength"]):
            errors.append(f"{path} is shorter than minLength")
        if "maxLength" in schema and len(value) > int(schema["maxLength"]):
            errors.append(f"{path} is longer than maxLength")
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        if "minimum" in schema and value < schema["minimum"]:
            errors.append(f"{path} is below minimum")
        if "maximum" in schema and value > schema["maximum"]:
            errors.append(f"{path} is above maximum")
    if isinstance(value, list):
        if "minItems" in schema and len(value) < int(schema["minItems"]):
            errors.append(f"{path} has too few items")
        if "maxItems" in schema and len(value) > int(schema["maxItems"]):
            errors.append(f"{path} has too many items")
        item_schema = schema.get("items")
        if isinstance(item_schema, dict):
            for index, item in enumerate(value):
                errors.extend(validate(item, item_schema, f"{path}[{index}]"))
    if isinstance(value, dict):
        props_raw = schema.get("properties")
        props = props_raw if isinstance(props_raw, dict) else {}
        required_raw = schema.get("required")
        required = required_raw if isinstance(required_raw, list) else []
        for key in required:
            if key not in value:
                errors.append(f"{path}.{key} is required")
        additional = schema.get("additionalProperties", True)
        for key, item in value.items():
            if key in props:
                errors.extend(validate(item, props[key], f"{path}.{key}"))
            elif additional is False:
                errors.append(f"{path}.{key} is not allowed")
            elif isinstance(additional, dict):
                errors.extend(validate(item, additional, f"{path}.{key}"))
    return errors


def _is_type(value: Any, kind: str) -> bool:
    if kind == "object":
        return isinstance(value, dict)
    if kind == "array":
        return isinstance(value, list)
    if kind == "string":
        return isinstance(value, str)
    if kind == "integer":
        return isinstance(value, int) and not isinstance(value, bool)
    if kind == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if kind == "boolean":
        return isinstance(value, bool)
    if kind == "null":
        return value is None
    return False
